### PR TITLE
Add types for Twitter-Autohook

### PR DESCRIPTION
**Pull request by dubisdev**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an npm package, match the name. If not, do not conflict with the name of an npm package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] Represents shape of module/library [correctly](https://www.typescriptlang.org/docs/handbook/declaration-files/library-structures.html)
- [x] `tslint.json` [should contain](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#linter-tslintjson) `{ "extends": "@definitelytyped/dtslint/dt.json" }`, and no additional rules.
- [x] `tsconfig.json` [should have](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#tsconfigjson) `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

File: types/twitter-autohook/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+// Project: https://github.com/twitterdev/autohook#readme
+// Definitions by: David Jiménez <https://github.com/dubisdev>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+
+/// <reference types="node" />
+
+import EventEmitter = require('events');
+
+export class Autohook extends EventEmitter {
+    constructor(configuration: {
```
Comment (peterblazejewicz) [inline types/twitter-autohook/index.d.ts:11]: please review the configuration, it looks like being completely optional with all properties optional as well (with defualts):
https://github.com/twitterdev/autohook/blob/master/index.js#L118-L126

File: types/twitter-autohook/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+        consumer_key: string;
+        consumer_secret: string;
+        ngrok_secret?: string;
+        env: string;
+        port: number;
+        headers?: any[];
+    });
+
+    startServer(): void;
+    setWebhook(webhookUrl: string): Promise<unknown>;
```
Comment (peterblazejewicz) [inline types/twitter-autohook/index.d.ts:23]:
```suggestion
    /** @async */
    setWebhook(webhookUrl: string): Promise<unknown>;
```
https://jsdoc.app/tags-async.html

File: types/twitter-autohook/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+    getWebhooks(): Promise<unknown>;
```
Comment (peterblazejewicz) [inline types/twitter-autohook/index.d.ts:24]:
```suggestion
    /** @async */
    getWebhooks(): Promise<unknown>;
```

File: types/twitter-autohook/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+    removeWebhook(webhook: unknown): Promise<void>;
```
Comment (peterblazejewicz) [inline types/twitter-autohook/index.d.ts:25]:
```suggestion
    /** @async */
    removeWebhook(webhook: unknown): Promise<void>;
```

File: types/twitter-autohook/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    removeWebhooks(): Promise<void>;
```
Comment (peterblazejewicz) [inline types/twitter-autohook/index.d.ts:26]:
```suggestion
    /** @async */
    removeWebhooks(): Promise<void>;
```

File: types/twitter-autohook/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+    start(webhookUrl?: string): Promise<void>;
```
Comment (peterblazejewicz) [inline types/twitter-autohook/index.d.ts:27]:
```suggestion
    /** @async */
    start(webhookUrl?: string): Promise<void>;
```

File: types/twitter-autohook/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+    subscribe(options: { oauth_token: string; oauth_token_secret: string; screen_name?: string }): Promise<true>;
```
Comment (peterblazejewicz) [inline types/twitter-autohook/index.d.ts:28]:
```suggestion
    /** @async */
    subscribe(options: { oauth_token: string; oauth_token_secret: string; screen_name?: string }): Promise<true>;
```

File: types/twitter-autohook/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+    unsubscribe(userId: string): Promise<true>;
```
Comment (peterblazejewicz) [inline types/twitter-autohook/index.d.ts:29]:
```suggestion
    /** @async */
    unsubscribe(userId: string): Promise<true>;
```

File: types/twitter-autohook/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+}
+
+export class WebhookURIError extends TwitterError {}
+
+export class TwitterError extends Error {
```
Comment (peterblazejewicz) [inline types/twitter-autohook/index.d.ts:34]: where does this symbols come from in the source?
https://github.com/twitterdev/autohook/blob/master/index.js#L355-L361

Comment (peterblazejewicz) [reply types/twitter-autohook/index.d.ts:34]: IMO it's internal, not exported

File: types/twitter-autohook/index.d.ts (hunk 1 above)
Comment (dubisdev) [reply types/twitter-autohook/index.d.ts:11]: Hi, I think they left defaults to ensure the `trim` method, but then the constructor throws error if they are empty, so I think they shouldn't be optional.

https://github.com/twitterdev/autohook/blob/master/index.js#L128-L133

Comment (peterblazejewicz) [reply types/twitter-autohook/index.d.ts:11]: Yes, but I think that means all setup is required to be from environment, say on AZ devops setup, note that they are checking if only ENV vars are set, e.g. 'headers' is not validated at runtime, can be defaulted

Comment (dubisdev) [reply types/twitter-autohook/index.d.ts]: It's true, you're right. It's already solved, thanks for your review and for the work all your team does 🙌

**Review by peterblazejewicz — APPROVED** [review]
LGTM!
@dubisdev thanks!

**Comment by dubisdev** [comment]
Ready to merge

**Pull request merged**